CCDB-3949: remove problematic dependencies that have CVE issues

ncliang opened the pull request:
## Type of PR

- [ ] Documentation changes
- [x] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR
As described in #380 , there are a number of dependencies pulled in transitively that have CVE issues discovered in Twistlock. This PR aims to remove these dependencies by:
- Bumping `azure-cosmos` version to latest version
- Bumping `azure-core-http-netty` version to latest version
- Explicitly replacing `log4j:log4j:1.2.17` with `io.confluent:confluent-log4j:1.2.17-cp2`

The twistlock scan after applying this PR contains no critical or high severity CVEs
![Screen Shot 2021-05-13 at 1 02 38 PM](https://user-images.githubusercontent.com/816633/118181290-6e5a5680-b3ec-11eb-84ee-cd1562733b90.png)


## Observability + Testing
- What changes or considerations did you make in relation to observability?
N/A

- Did you add testing to account for any new or changed work?
Verified CVE fixes in Twistlock. Sanity check the connector after applying this PR with scripts in https://github.com/vdesabou/kafka-docker-playground/tree/master/connect/connect-azure-cosmosdb-sink

## Review notes
The change includes changes from #382 . Please focus this review on changes in `pom.xml` only. I will refresh this PR once #382 has been merged.

## Issues Closed or Referenced
Closes #380

kushagraThapar commented on pom.xml:
```diff
@@ -42,21 +43,17 @@
-            <groupId>io.confluent</groupId>
-            <artifactId>kafka-schema-registry-maven-plugin</artifactId>
-            <version>6.0.1</version>
+            <groupId>com.azure</groupId>
```
`azure-cosmos` already brings `azure-core-http-netty` - why do we need this explicitly ?

ryancrawcour replied:
maybe he's bringing in a specific version of netty?

ncliang replied:
Yes, with the latest `azure-cosmos` , we transitively bring in `azure-core-http-netty`, which transitively pulls in `netty:4.1.60.Final`, which contains a critical CVE. I had to pull this in separately, pinned to a newer version to workaround.

![Screen Shot 2021-05-13 at 7 51 22 PM](https://user-images.githubusercontent.com/816633/118214496-aed5c680-b424-11eb-8f21-fa1c3b19b26f.png)

kushagraThapar replied:
@ncliang - we recently released azure-cosmos 4.15.0 - which transitively brings netty version 4.1.63.Final. Let's upgrade azure-cosmos to 4.15.0

ryancrawcour replied:
> we transitively bring in `azure-core-http-netty`, which transitively pulls in **`netty:4.1.60.Final`, which contains a critical CVE**

@kushagraThapar i believe that what @ncliang  is saying is that netty 4.1.63.Final has issues with it, so you should update the cosmos client to use a later version of netty (which he does here by explicitly setting the version he wants)

kushagraThapar replied:
@ryancrawcour now really, @ncliang is talking about version 4.1.60.Final and not 4.1.63.Final. 
And azure-cosmos version 4.15.0 actually depends on azure-core-http-netty version 1.9.2 - please refer here - https://mvnrepository.com/artifact/com.azure/azure-cosmos/4.15.0

@ncliang - please update azure-cosmos to 4.15.0 in this PR, and remove the azure-core-http-netty dependency from pom.xml

moderakh replied:
+1, as there are different netty component, including boring-ssl, and their versions need to match, I would just rely on Cosmos DB to bring its dependency. Please don't override.

moderakh replied:
@ncliang could you please take a look at the suggestion from @kushagraThapar

kushagraThapar reviewed (changes requested):
Thanks @ncliang for the contribution. Overall looks good to me, except one comment.

ryancrawcour commented on src/main/java/com/azure/cosmos/kafka/connect/sink/CosmosDBSinkConfig.java:
```diff
@@ -15,54 +15,11 @@
 public class CosmosDBSinkConfig extends CosmosDBConfig {
-
-    static final String COSMOS_USE_UPSERT_CONF = "connect.cosmos.sink.useUpsert";
-    private static final String COSMOS_USE_UPSERT_DEFAULT = "false";
```
these changes are part of another PR, right?

ncliang replied:
I ended up stacking this on #382 as the separation was not entirely clean. Part of the motivation for removing the schema-registry dependency was that it was pulling in Jersey, which is both unnecessary and contained a CVE.

ryancrawcour replied:
does that make the other PR redundant now?

ryancrawcour commented on src/main/java/com/azure/cosmos/kafka/connect/sink/CosmosDBSinkTask.java:
```diff
@@ -90,11 +90,7 @@ public void put(Collection<SinkRecord> records) {
-        } else {
-            container.createItem(recordValue);
-        }
+        container.upsertItem(recordValue);
```
@kushagraThapar what are the perf implications of always doing upsertItem instead of insertItem?

kushagraThapar replied:
Upsert internally does a read first, so there will be some performance degradation and more RU cost for each operation. However, this will make sure the correctness of the SDK in every case. That is basically the trade off we have decided at this point of time. 
Once we have covered all the corner cases of auto-generation of ids, we can then start using insert and expose the configuration back again to the customers.

kushagraThapar replied:
@moderakh - thoughts ?

ryancrawcour replied:
as @ncliang explained i believe this is an acceptable compromise short term. once we implement finer grained control over the checkpointing and have DLQ implemented properly we can look to reintroduce this. so i am ok with this as an interim.

moderakh replied:
upsert is the default on spark. and the perf difference is negligible. As the create story is not compatible with retry we should remove this for now.

ryancrawcour reviewed (dismissed):
LGTM. happy to merge as-is.

moderakh reviewed (dismissed):
@ncliang thanks for the PR.

Please address @kushagraThapar's comment on upgrade to the latest released version of azure-cosmos and avoid overriding transitive dependencies of azure-cosmos

ncliang commented:
Thanks for review @kushagraThapar , @ryancrawcour , @moderakh . I have rebased the PR after merge and updated the POM to use `azure-core:4.15.0` . When I made the PR, I only saw a couple beta builds for `azure-core:4.15.0` , hence the workaround. This should look much better now. PTAL?
